docs/src/ref/mcmc.md

```diff
@@ -201,6 +201,207 @@ new_trace = permute_move(trace, 2)
+        m2 = ({:m2} ~ gamma(1, 1))
+    else
+        m = ({:m} ~ gamma(1, 1))
+        (m1, m2) = (m, m)
+    end
+    {:y1} ~ normal(m1, 0.1)
+    {:y2} ~ normal(m2, 0.1)
+end
+```
+Because this model has stochastic control flow, it represents two distinct structural hypotheses about how the observed data could have been generated:
+If `:z` is `true` then we enter the first branch, and we hypothesize that the two data points were generated from separate means, sampled at addresses `:m1` and `:m2`.
+If `:z` is `false` then we enter the second branch, and we hypohesize that there is a single mean that explains both data points, sampled at address `:m`.
+
+We want to construct an MCMC kernel that is able to transition between these two distinct structural hypotheses.
+We could construct such a kernel with the simpler 'selection' variant of Metropolis-Hastings, by selecting the address `:z`, e.g.:
+```julia
+select_mh_structure_kernel(trace) = mh(trace, select(:z))[1]
+```
+Sometimes, this kernel would propose to change the value of `:z`.
+We could interleave this kernel with another kernel that does inference over the mean random choices, without changing the structure, e.g.:
+```julia
+@gen function fixed_structure_proposal(trace)
+    if trace[:z]
+        {:m1} ~ normal(trace[:m1], 0.1)
+        {:m2} ~ normal(trace[:m2], 0.1)
+    else
+        {:m} ~ normal(trace[:m], 0.1)
+    end
+end
+
+fixed_structure_kernel(trace) = mh(trace, fixed_structure_proposal, ())[1]
+```
+Combining these together, and applying to particular data and with a specific initial hypotheses:
+```julia
+(y1, y2) = (1.0, 1.3)
+trace, = generate(model, (), choicemap((:y1, y1), (:y2, y2), (:z, false), (:m, 1.2)))
+for iter=1:100
+    trace = select_mh_structure_kernel(trace)
+    trace = fixed_structure_kernel(trace)
+end
+```
+However, this algorithm will not be very efficient, because the internal proposal used by the selection variant of MH is not specialized to the model.
+In particular, when switching from the model with a single mean to the model with two means, the values of the new addresses `:m1` and `:m2` will be proposed from the prior distribution.
+This is wasteful, since if we have inferred an accurate value for `:m`, we expect the values for `:m1` and `:m2` to be near this value.
+The same is true when proposing a structure change in the opposite direction.
+That means it will take many more steps to get an accurate estimate of the posterior probability distribution on the two structures.
+
+We would like to use inferred values for `:m1` and `:m2` to inform our proposal for the value of `:m`.
+For example, we could take the geometric mean:
+```julia
+m = sqrt(m1 * m2)
+```
+However, there are many combinations of `m1` and `m2` that have the same geometric mean.
+In other words, the geometric mean is not *invertible*.
+However, if we return the additional degree of freedom alongside the geometric mean (`dof`), then we do have an invertible function:
```

Comment:
> @marcoct, to help me review further, could you please explain a bit what the "first pass" and the "Jacobian pass" are, and how they differ?

@bzinberg The first pass actually applies the involution (and now produces the new trace as well), but it doesn't compute the Jacobian correction. The second pass computes the reduced Jacobian matrix. We have to run the second pass separately because (i) we don't know what addresses actually need to be included in the reduced Jacobian until after we have run the involution, and (ii) because the AD system being used (ForwardDiff) wants to compute Jacobians for functions that map arrays to arrays.